proposals/0000-linear-types.rst

```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+consumption of the result of the function was never complete, so the
+consumption of the argument need not be either. However, because
+exceptions can be caught, a program can observe a state where a value
+``v`` has been passed to a linear function ``f`` but the call ``f v``
+has exited (with an exception) without consuming ``v``. So while, the
+guarantee provided by linear functions holds for converging
+computations, we must weaken it in case of divergence:
+
+- Attempting to consume exactly once ``f v``, when ``f`` is a linear
+  function, will consume ``v`` exactly once if the consumption of ``f
+  v`` converges, and *at most once* if it diverges.
+
+Where "consuming at most once" is defined by induction, like
+"consuming exactly once", but every sub-consumption is optional.
+
+In the paper, we gave a simplified specification of a linear ``IO``
+monad (called ``IOL``) which ignored the issue of exception for the
+sake of simplicity. Can we, still, write a resource-safe ``RIO`` monad
+with linear types despite the added difficulty of exceptions? Yes, as
+this section will show.
+
+Concretely, how do we ensure that the sockets from the example API are
+always closed, even in presence of exceptions? This boils down to how
+the ``RIO`` monad is implemented. Below is a sketch of one possible
+implementation of ``RIO`` (see `here
+<https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs>`_
+for a detailed implementation).
+
+First, note that since Haskell program are of type ``IO ()``, we need a
+way to run ``RIO`` in an ``IO`` computation, this is provided by the
+function
+
+::
+
+  runRIO :: RIO (Unrestricted a) -> IO a
+
+In order to achieve resource safety in presence of exception, ``runRIO``
+is tasked with releasing any live resource in case of exception.
+
+To implement this, ``RIO`` keeps a table of release actions, to be used
+in case of exceptions. Each resource implemented in the ``RIO``
+abstraction registers a release action in the release action table
+when they are acquired.
+
+If no exception occurs, then all resources have been released by the
+program. In case an exception occurs, the program jumps to ``runRIO``,
+which releases the leftover resources.
+
+An alternative strategy would be to add terminators on every resources
+acquired in ``RIO``. Release in the non-exceptional case would still
+be performed by the program, and the GC would be responsible for
+releasing resources in case of exception. The release in case of
+exception would be, however, less timely.
+
+Can ``RIO`` have a ``catch``?
+=============================
+
+It is possible to catch exceptions inside of ``RIO``. But in order to
+ensure resource safety, the type cannot be linear:
+
+::
+
+  catchL :: Exception e
+         => RIO (Unrestricted a)
+	 -> (e -> RIO (Unrestricted a))
+	 -> RIO (Unrestricted a)
+
+That is: no linear resource previously allocated can be referenced in
```

Comment:
As far as I can see, you can make the body linear. And so, the linear variables can occur in it. If a variable turns out not to be used before an exception occurs it can be freed with the mechanism described above. Am I missing something?

Reply:
The problem here is that you have to know how to free the linear thing. Which you usually do not. The mechanism described above works because there is a cooperation between `RIO` and the resource types it manages. It can't, in general, be assumed of linear variable that they cooperate this way.

Having a linear catch this way would make it possible to break abstractions enforced by linear types.

Reply:
Alright, I was assuming that all linear resources were allocated using RIO. 

Perhaps you could add a line saying "because there could be linear resources coming from outside RIO".

Reply:
The truth is, though, that even if the linear resources are allocated by RIO, we would still need to figure out, in implementation of `catch`, which resources were captured by the body and must be deallocated by `catch` in case of exception. I don't know how to do this either. Otherwise, the resource would be held dangling until the entire `runRIO` scope is finished. Which may be a long time.

But this is less of a logic problem and more of an implementation detail. I suppose.

Reply:
I thought that this is what RIO does (closing resources only when the entire RIO finishes).

Reply:
Well, not quite: they are closed when a deallocating primitive is called. But if there is an exception, the entire `RIO` aborts, and frees remaining resources.

So the point is that you tend to have one big `runRIO` scope, and you do your resource management in it using linear types.

But if we start catching exceptions carelessly in the middle of an `RIO`, we may lose the ability to manage some resources, effectively leaking resources.